### PR TITLE
GH-100242: bring functools.py partial implementation more in line with C code

### PR DESCRIPTION
**Pull request by cfbolz**
in `partial.__new__`, before checking for the existence of the attribute 'func', first check whether the argument is an instance of partial.

* Issue: gh-100242

### CODE REVIEW

File: Lib/functools.py [hunk 1]
```diff
@@ -284,7 +284,7 @@ def __new__(cls, func, /, *args, **keywords):
         if not callable(func):
             raise TypeError("the first argument must be callable")
 
-        if hasattr(func, "func"):
+        if isinstance(func, partial) and hasattr(func, "func"):
```
Comment (AlexWaygood) [inline Lib/functools.py]: Will an instance of `partial` ever not have a `func` attribute? Could this be simplified to just an `isinstance()` check?

```suggestion
        if isinstance(func, partial):
```

Comment (cfbolz) [reply Lib/functools.py]: yes, that works for me too. (you can certainly remove the `func` attribute from a pure python `partial` instance, but the resulting object is not very useful)

Comment (cfbolz) [reply Lib/functools.py]: if we do that, maybe it's also a good idea to just use an isinstance check in the C implementation as well? then the two approaches are straightforwardly identical.

Comment (AlexWaygood) [reply Lib/functools.py]: > you can certainly remove the func attribute from a pure python partial instance, but the resulting object is not very useful

Yeah, if people are deleting the `func` attribute from `partial` objects, I have a lot of questions about their broader approach to programming in general :)

> maybe it's also a good idea to just use an isinstance check in the C implementation as well?

I'm not qualified to comment on the C implementation, so I'll defer to you and @rhettinger on that question :)

Comment (cfbolz) [reply Lib/functools.py]: > Yeah, if people are deleting the func attribute from partial objects, I have a lot of questions about their broader approach to programming in general :)

this argument convinced me ;-). I just went with the `isinstance` check in both places.

**Comment by rhettinger** [comment]
As a first approximation this looks good.  In the next couple of days, I can devote some time to give it more thought.

**Comment by cfbolz** [comment]
@rhettinger very gentle ping, would you maybe have some time to look into this PR?

**Review by ambv — APPROVED** [review]
The only risk I'm seeing here is the subtle change that now allows a subclass of `functools.partial` with an overriden `__call__` whereas this would be rejected before in the C version.

As long as we accept this change for 3.13 alone, this is uncontroversial. As for whether this should land in 3.12.4, I'm leaving this question open. My intuition says it's fine.

**Pull request merged**